include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1143,27 +1144,113 @@ struct __early_exit_find_or
 // parallel_find_or - sync pattern
 //------------------------------------------------------------------------
 
+template <typename Tag>
+struct __parallel_find_or_n_groups_tuner
+{
+    // Calculate the number of work groups.
+    //
+    // @param _ExecutionPolicy&& __exec - policy
+    // @param std::size_t __n_groups - the source amount of work-groups
+    // @param std::size_t __wgroup_size - the source work-group size
+    // @param std::size_t __rng_n - the source data size
+    // @return std::size_t - evaluated number of work-groups
+    template <typename _ExecutionPolicy>
+    std::size_t
+    operator()(_ExecutionPolicy&& /*__exec*/, std::size_t __n_groups, const std::size_t /*__wgroup_size*/,
+               const std::size_t /*__rng_n*/) const
+    {
+        return __n_groups;
+    }
+};
+
+template <>
+struct __parallel_find_or_n_groups_tuner<oneapi::dpl::__internal::__device_backend_tag>
+{
+    // Calculate the number of work groups.
+    //
+    // @param _ExecutionPolicy&& __exec - policy
+    // @param std::size_t __n_groups - the source amount of work-groups
+    // @param std::size_t __wgroup_size - the source work-group size
+    // @param std::size_t __rng_n - the source data size
+    // @return std::size_t - evaluated number of work-groups
+    template <typename _ExecutionPolicy>
+    std::size_t
+    operator()(_ExecutionPolicy&& /*__exec*/, std::size_t __n_groups, const std::size_t __wgroup_size,
+               const std::size_t __rng_n) const
+    {
+        // No tuning for FPGA_EMU
+#if _ONEDPL_FPGA_EMU
+        return __n_groups;
+#endif
+
+        // If all source data fits into one work-group, then we need only one work-group
+        if (__rng_n <= __wgroup_size)
+            return 1;
+
+        // Size: [ 16'777'216, ..,  67'108'864 ) -> minimum number of iterations per work-item is 164
+        // Size: [  4'194'304, ..,  16'777'216 ) -> minimum number of iterations per work-item is  32
+        // Size: [  1'048'576, ...,  4'194'304 ) -> minimum number of iterations per work-item is  16
+        // Size: [    262'144, ...,  1'048'576 ) -> minimum number of iterations per work-item is   8
+        if (__n_groups > 1)
+        {
+            constexpr std::array<std::size_t, 6> __lower_bounds_of_sizes         = { 262'144, 1'048'576, 4'194'304, 16'777'216, 67'108'864 };
```

Comment:
I think I follow this approach but we might want to simplify this. The current values could be derived from the input range since we double `__required_iters_per_work_items ` for multiples of four of `__lower_bounds_of_sizes`. Calculating them could remove the need for `find_if`.

Reply:
I think my approach is more customizable in the future: the current values is only some best values found by me on concrete device. In the future we may correct them.